feat: implement customizable prompt for agents

karindralinux opened the pull request:
- Added a new `prompt.ts` file containing interfaces and default prompt functions for various agent types, including Task Breakdown, Task Replanning, Action Agent, Summarizer Agent, and Enhance Prompt.
- Introduced a `getPrompt` helper function in `agents.ts` to retrieve custom prompts or fallback to defaults.
- Updated `EnhancePromptAgent`, `TaskBreakdownAgent`, `TaskReplanningAgent`, and `ActionAgent` classes to utilize the new prompt system, enhancing modularity and reusability.
- Enhanced `ReactAgentBuilder` to support custom prompts configuration and retrieval.
- Added example usage in `testCustomAgentPrompt.ts` to demonstrate the new functionality.

This update improves the flexibility and maintainability of agent prompts, allowing for easier customization and better adherence to DRY principles.

karindralinux committed:
feat: implement customizable prompt for agents

- Added a new `prompt.ts` file containing interfaces and default prompt functions for various agent types, including Task Breakdown, Task Replanning, Action Agent, Summarizer Agent, and Enhance Prompt.
- Introduced a `getPrompt` helper function in `agents.ts` to retrieve custom prompts or fallback to defaults.
- Updated `EnhancePromptAgent`, `TaskBreakdownAgent`, `TaskReplanningAgent`, and `ActionAgent` classes to utilize the new prompt system, enhancing modularity and reusability.
- Enhanced `ReactAgentBuilder` to support custom prompts configuration and retrieval.
- Added example usage in `testCustomAgentPrompt.ts` to demonstrate the new functionality.

This update improves the flexibility and maintainability of agent prompts, allowing for easier customization and better adherence to DRY principles.

ans-4175 commented on core/agents.ts:
```diff
@@ -50,6 +60,16 @@ The above documents have been loaded and their data is available for analysis. Y
 
+
+// Helper function to get custom prompt or default
+function getPrompt<T>(
+  config: Record<string, any>, 
+  agentType: keyof AgentPrompts
+): ((params: T) => string) {
+  const customPrompts = config?.configurable?.agentConfig?.prompts as AgentPrompts;
+  return customPrompts?.[agentType] as ((params: T) => string) || DEFAULT_PROMPTS[agentType];
+}
```
please check this @karindralinux

ans-4175 commented on core/agents.ts:
```diff
@@ -157,63 +169,32 @@ export class TaskBreakdownAgent extends BaseAgent {
-    
-    // Priority 1 & 2: Handle completion scenarios (existing summarize task OR objective achieved)
-    const hasSummarizeTask = currentTask && currentTask.toLowerCase().includes("summarize");
+    // ... existing logic for early returns ...
```
logExecution jangan dibuang @karindralinux 
Sama bagian yang terkait getCurrent task dll juga

ans-4175 commented on core/agents.ts:
```diff
@@ -157,63 +169,32 @@ export class TaskBreakdownAgent extends BaseAgent {
-    if (hasSummarizeTask || state.objectiveAchieved) {
-      if (hasSummarizeTask) {
-        // Already has summarize task - pass through
-        TaskReplanningAgent.logExecution("TaskReplanningAgent", "summarizeTaskDetected", {
```
sama bagian-bagian yang ada logexecution jangan dibuang juga @karindralinux

ans-4175 commented on core/index.ts:
```diff
@@ -36,6 +37,12 @@ export interface ReactAgentConfig {
+    taskBreakdown?: (params: TaskBreakdownParams) => string;
+    taskReplanning?: (params: TaskReplanningParams) => string;
+    summarizerAgent?: (params: SummarizerAgentParams) => string;
+    actionAgent?: (params: ActionAgentParams) => string;
```
iyah enhancePrompt, ga dikirim dari config depan @karindralinux ?

ans-4175 commented on core/index.ts:
```diff
@@ -547,6 +554,25 @@ class ReactAgentBuilder {
+    return DEFAULT_PROMPTS[agentType].toString();
+  }
+
+  public getAllDefaultPrompts(): { [key: string]: string }[] {
+    return Object.keys(DEFAULT_PROMPTS).map((key: string ) => ({
+      [key]: DEFAULT_PROMPTS[key as keyof typeof DEFAULT_PROMPTS].toString(),
+    }));
```
Coba cek lagi @karindralinux

ans-4175 commented on core/index.ts:
```diff
@@ -547,6 +554,25 @@ class ReactAgentBuilder {
+    return DEFAULT_PROMPTS[agentType].toString();
+  }
+
+  public getAllDefaultPrompts(): { [key: string]: string }[] {
+    return Object.keys(DEFAULT_PROMPTS).map((key: string ) => ({
+      [key]: DEFAULT_PROMPTS[key as keyof typeof DEFAULT_PROMPTS].toString(),
+    }));
+  }
```
sama coba cek lagi @karindralinux

ans-4175 commented on example/testCustomAgentPrompt.ts:
selain testCustomAgentPromp, mungkin sekalian bikin docs-nya buat di webdocs, dan references readme @karindralinux

ans-4175 commented on example/testCustomAgentPrompt.ts:
```diff
@@ -0,0 +1,112 @@
+- Job market insights`;
+            },
+
+            summarizerAgent: (params: SummarizerAgentParams) => {
```
terutama docs terkait gimana bikin prompt template dan consider params yang relevan

ans-4175 reviewed (changes requested):
Coba cek lagi @karindralinux

karindralinux committed:
fix: restore log execution on task replanning agent

karindralinux committed:
chore: add custom enhance prompt on agent config

karindralinux committed:
docs: add docs for custom prompt

- Introduced a comprehensive system for customizing agent behavior through tailored prompts, allowing for domain-specific and industry-focused functionalities.
- Updated the README to include a new section on Custom Agent Prompts, detailing key features, available agent types, and use cases.
- Added a new documentation file for Custom Agent Prompts, providing in-depth guidance and examples for implementation.
- Refactored the `getAllDefaultPrompts` method for improved readability and efficiency.

This update significantly enhances the flexibility of agent configurations, enabling more specialized and effective interactions.

karindralinux committed:
Merge branch 'main' of github.com:Delegasi-Tech/DelReact into feat/agent-custom-prompt-config

ans-4175 reviewed (approved):
Minor aja catatan yang ada tadi di komen @karindralinux , habis itu kayaknya oke ya?

karindralinux commented on core/index.ts:
```diff
@@ -751,6 +759,25 @@ class ReactAgentBuilder {
+      taskBreakdown : this.config.prompts?.taskBreakdown?.toString() || DEFAULT_PROMPTS.taskBreakdown.toString(),
+      taskReplanning : this.config.prompts?.taskReplanning?.toString() || DEFAULT_PROMPTS.taskReplanning.toString(),
+      actionAgent : this.config.prompts?.actionAgent?.toString() || DEFAULT_PROMPTS.actionAgent.toString(),
+      summarizerAgent : this.config.prompts?.summarizerAgent?.toString() || DEFAULT_PROMPTS.summarizerAgent.toString(),
```
oiya kelewat mas

karindralinux committed:
refactor: update prompt handling in agent configuration

- Modified the `getPrompt` function to return either a string or a function, enhancing flexibility in prompt retrieval.
- Updated the `ReactAgentBuilder` to include the `enhancePrompt` in the configuration, ensuring it is properly handled alongside other prompts.

These changes improve the customization capabilities of agent prompts, allowing for more dynamic interactions.

ans-4175 reviewed (approved):
Alrite, LGTM
